lang/machine.md

```diff
@@ -136,6 +123,112 @@ We also define some helper functions that will be useful later.
 
 ```rust
 impl<M: Memory> Machine<M> {
+    fn cur_frame(&self) -> StackFrame<M> {
+        let Some(active_thread) = self.thread_manager.active_thread else {
+            panic!("`cur_frame` called without active thread!");
+        };
+
+        self.thread_manager.threads[active_thread].cur_frame()
+    }
+
+    fn mutate_cur_frame<O>(&mut self, f: impl FnOnce(&mut StackFrame<M>) -> O) -> O {
+        let Some(active_thread) = self.thread_manager.active_thread else {
+            panic!("`mutate_cur_frame` called without active thread!");
+        };
+
+        self.thread_manager.threads.mutate_at(active_thread, |thread| thread.mutate_cur_frame(f))
+    }
+
+    fn mutate_cur_stack<O>(&mut self, f: impl FnOnce(&mut List<StackFrame<M>>) -> O) -> O {
+        let Some(active_thread) = self.thread_manager.active_thread else {
+            panic!("`cur_stack` called without active thread!");
+        };
+
+        self.thread_manager.threads.mutate_at(active_thread, |thread| f(&mut thread.stack))
+    }
+
+    fn cur_thread(&self) -> Thread<M> {
+        let Some(active_thread) = self.thread_manager.active_thread else {
+            panic!("`cur_thread` called without active thread!");
+        };
+
+        self.thread_manager.threads[active_thread]
+    }
+
+    fn mutate_cur_thread<O>(&mut self, f: impl FnOnce(&mut Thread<M>) -> O) -> O {
+        let Some(active_thread) = self.thread_manager.active_thread else {
+            panic!("`cur_thread` called without active thread!");
+        };
+
+        self.thread_manager.threads.mutate_at(active_thread, f)
+    } 
+}
+
+impl<M: Memory> StackFrame<M> {
+    /// jump to the beginning of the given block.
+    fn jump_to_block(&mut self, b: BbName) {
+        self.next_block = b;
+        self.next_stmt = Int::ZERO;
+    }
+}
+```
+
+## MiniRust Thread
+
+This defines the state that makes up a thread of execution in MiniRust. It is used by the Abstract Machine to make progress on.
+
+```rust
+// This type contains all information the abstract machine needs
+// to continue a thread of execution.
+pub struct Thread<M: Memory> {
+    state: ThreadState,
+
+    // The stack.
+    stack: List<StackFrame<M>>,
+}
+
+pub enum ThreadState {
+    // The thread is enabled and can get executed.
+    Enabled,
+    // The thread is trying to join another thread and is blocked until that thread finishes.
+    BlockedOnJoin(ThreadId),
+    // The thread has terminated.
+    Terminated,
+}
+
+type ThreadId = Int;
```

Comment:
Please keep all the type definitions together at the top of the file. They will be the most common thing people are looking for.